src/lru-store.ts

```diff
@@ -1,39 +1,55 @@
-import type { AsyncStore } from 'zarr/types/storage/types';
+import type { Readable } from '@zarrita/storage';
 import QuickLRU from 'quick-lru';
 
-export class LRUCacheStore<S extends AsyncStore<ArrayBuffer>> {
-  cache: QuickLRU<string, Promise<ArrayBuffer>>;
+type RangeQuery =
+  | {
+      offset: number;
+      length: number;
+    }
+  | {
+      suffixLength: number;
+    };
 
-  constructor(public store: S, maxSize: number = 100) {
-    this.cache = new QuickLRU({ maxSize });
-  }
+function normalizeKey(key: string, range?: RangeQuery) {
+  if (!range) return key;
+  if ('suffixLength' in range) return `${key}:-${range.suffixLength}`;
+  return `${key}:${range.offset}:${range.offset + range.length - 1}`;
+}
 
-  getItem(...args: Parameters<S['getItem']>) {
+export function lru<S extends Readable>(store: S, maxSize: number = 100) {
+  const cache = new QuickLRU<string, Promise<Uint8Array | undefined>>({ maxSize });
+  let getRange = store.getRange ? store.getRange.bind(store) : undefined;
+  function get(...args: Parameters<S['get']>) {
     const [key, opts] = args;
-    if (this.cache.has(key)) {
-      return this.cache.get(key)!;
-    }
-    const value = this.store.getItem(key, opts).catch((err) => {
-      this.cache.delete(key);
+    const cacheKey = normalizeKey(key);
+    const cached = cache.get(cacheKey);
+    if (cached) return cached;
+    const result = Promise.resolve(store.get(key, opts)).catch((err) => {
+      cache.delete(cacheKey);
       throw err;
     });
-    this.cache.set(key, value);
-    return value;
-  }
-
-  async containsItem(key: string) {
-    return this.cache.has(key) || this.store.containsItem(key);
+    cache.set(cacheKey, result);
+    return result;
   }
-
-  keys() {
-    return this.store.keys();
-  }
-
-  deleteItem(key: string): never {
-    throw new Error('deleteItem not implemented');
-  }
-
-  setItem(key: string, value: ArrayBuffer): never {
-    throw new Error('setItem not implemented');
+  if (getRange) {
+    getRange = (...args: Parameters<NonNullable<S['getRange']>>) => {
+      const [key, range, opts] = args;
+      const cacheKey = normalizeKey(key, range);
+      const cached = cache.get(cacheKey);
+      if (cached) return cached;
+      const result = Promise.resolve(getRange!(key, range, opts)).catch((err) => {
```

Comment:
```suggestion
  if (getRange) {
    const _getRange = getRange;
    getRange = (...args: Parameters<NonNullable<S['getRange']>>) => {
      const [key, range, opts] = args;
      const cacheKey = normalizeKey(key, range);
      const cached = cache.get(cacheKey);
      if (cached) return cached;
      const result = Promise.resolve(_getRange!(key, range, opts)).catch((err) => {
```

Reply:
Great catch. Thanks for digging into this.